tools/vm-migrator/migrator/instance.py

```diff
@@ -0,0 +1,397 @@
+                        return True
+        return False
+    except KeyError:
+        return False
+
+
+def get_node_group(instance):
+    if (is_hosted_on_sole_tenant(instance)):
+        return instance['scheduling']['nodeAffinities'][0]['values'][0]
+    return None
+
+
+def get_updated_node_group(node_group):
+    try:
+        if (node_group_mapping.FIND.get(node_group)):
+            config = {
+                "scheduling": {
+                    "nodeAffinities": [{
+                        "key": 'compute.googleapis.com/node-group-name',
+                        "operator": "IN",
+                        "values": [node_group_mapping.FIND[node_group]]
+                    }]
+                }
+            }
+            logging.info("Found a matching node group %s for %s" %
+                         (node_group, node_group_mapping.FIND.get(node_group)))
+            return config
+        else:
+            return None
+    except KeyError:
+        return None
+
+
+def parse_self_link(self_link):
+    if (self_link.startswith('projects')):
+        self_link = "/" + self_link
+    response = re.search(r"\/projects\/(.*?)\/zones\/(.*?)\/instances\/(.*?)$",
+                         self_link)
+    if (len(response.groups()) != 3):
+        raise Exception('Invalid SelfLink Format')
+    return {
+        'instance_id': response.group(3),
+        'zone': response.group(2),
+        'project': response.group(1)
+    }
+
+
+def shutdown_instance(compute, project, zone, instance_name):
+    result = compute.instances().stop(project=project,
+                                      zone=zone,
+                                      instance=instance_name).execute()
+    return result
+
+
+def shutdown(project, zone, instance_name):
+    try:
+        waited_time = rate_limit.wait()  # wait before starting the task
+        logging.info(f"  task: waited for {waited_time} secs")
+        compute = get_compute()
+        logging.info("Shutting Down Instance %s ", (instance_name))
+        result = shutdown_instance(compute, project, zone, instance_name)
+        wait_for_zonal_operation(compute, project, zone, result['name'])
+        return instance_name
+    except Exception as ex:
+        logging.error(ex)
+        print(ex)
+        raise ex
+
+
+def start(project, zone, instance_name):
+    try:
+        waited_time = rate_limit.wait()  # wait before starting the task
+        logging.info(f"  task: waited for {waited_time} secs")
+        compute = get_compute()
+        logging.info("Starting Instance %s ", (instance_name))
+        result = compute.instances().start(project=project,
+                                           zone=zone,
+                                           instance=instance_name).execute()
+        wait_for_zonal_operation(compute, project, zone, result['name'])
+        return instance_name
+    except Exception as ex:
+        logging.error(ex)
+        print(ex)
+        raise ex
+
```

Comment:
removed the print statement, but if i remove handling `Exception` i am not sure which custom exception to raise since the [documentation](https://googleapis.github.io/google-api-python-client/docs/dyn/compute_beta.instances.html#start) to use `compute.instances().start` does not mention any custom exception it raises hence handling everything and re-raising it to mark the executor as failed.